File: src/lab/prada/android/ui/infinitescroll/InfiniteScrollAdapter.java
package lab.prada.android.ui.infinitescroll;

import java.util.Vector;
import java.util.concurrent.atomic.AtomicInteger;

import android.content.Context;
import android.database.DataSetObserver;
import android.os.Looper;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Adapter;
import android.widget.BaseAdapter;
import android.widget.GridView;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;

public class InfiniteScrollAdapter<T extends BaseAdapter> extends BaseAdapter {

    private final T mAdapter;
    private final View mProgressView;
    private Vector<PullDownToRefreshListener> mListeners = new Vector<PullDownToRefreshListener>();
    
    private final static int NONE_STATE = 0;
    private final static int REFRESHING_STATE = 1;
    private AtomicInteger state = new AtomicInteger(NONE_STATE);
    private boolean mCanReadMore = true;

    public interface PullDownToRefreshListener {
        public void pullDownToRefresh();
    }

    public InfiniteScrollAdapter(Context context, T adapter, View progressView) {
        mAdapter = adapter;
        mProgressView = progressView;
    }

    public InfiniteScrollAdapter(Context context, T adapter, int itemWidth,
            int itemHeight) {
        mAdapter = adapter;
        RelativeLayout layout = new RelativeLayout(context);
        layout.setLayoutParams(new GridView.LayoutParams(itemWidth,
                itemHeight));
        layout.setGravity(Gravity.CENTER);
        layout.addView(new ProgressBar(context));

        mProgressView = layout;
    }

    public InfiniteScrollAdapter(T adapter, View progressView) {
        mAdapter = adapter;
        mProgressView = progressView;
    }

    @Override
    public boolean areAllItemsEnabled() {
        return mAdapter.areAllItemsEnabled();
    }

    @Override
    public View getDropDownView(int position, View convertView, ViewGroup parent) {
        return (isProgressViewPosition(position)) ? null : mAdapter
                .getDropDownView(position, convertView, parent);
    }

    @Override
    public int getItemViewType(int position) {
        return (isProgressViewPosition(position)) ? Adapter.IGNORE_ITEM_VIEW_TYPE
                : mAdapter.getItemViewType(position);
    }

    @Override
    public int getViewTypeCount() {
        return mAdapter.getViewTypeCount();
    }

    @Override
    public boolean hasStableIds() {
        return mAdapter.hasStableIds();
    }

    @Override
    public boolean isEmpty() {
        return mAdapter.isEmpty();
    }

    @Override
    public boolean isEnabled(int position) {
        return (isProgressViewPosition(position)) ? false : mAdapter
                .isEnabled(position);
    }

    @Override
    public void notifyDataSetChanged() {
        mAdapter.notifyDataSetChanged();
    }

    @Override
    public void notifyDataSetInvalidated() {
        mAdapter.notifyDataSetInvalidated();
    }

    @Override
    public void registerDataSetObserver(DataSetObserver observer) {
        mAdapter.registerDataSetObserver(observer);
    }

    @Override
    public void unregisterDataSetObserver(DataSetObserver observer) {
        mAdapter.unregisterDataSetObserver(observer);
    }

    @Override
    public int getCount() {
        return shouldShowProgressView() ? mAdapter.getCount() + 1 : mAdapter
                .getCount();
    }

    @Override
    public Object getItem(int position) {
        return (isProgressViewPosition(position)) ? null : mAdapter
                .getItem(position);
    }

    @Override
    public long getItemId(int position) {
        return (isProgressViewPosition(position)) ? -1 : mAdapter
                .getItemId(position);
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        if (isProgressViewPosition(position)) {
            if (isRefreshing() == false) {
                state.set(REFRESHING_STATE);
                if (mListeners != null) {
                    for (PullDownToRefreshListener listener : mListeners) {
                        listener.pullDownToRefresh();
                    }
                }
            }
            if (mProgressView.getVisibility() == View.GONE) {
                mProgressView.setVisibility(View.VISIBLE);
            }
            return mProgressView;
        } else {
            return mAdapter.getView(position, convertView, parent);
        }
    }

    public T getOriginalAdapter() {
        return mAdapter;
    }

    public void removeListener(PullDownToRefreshListener listener) {
        if (listener != null) {

        }
    }

    public void addListener(PullDownToRefreshListener listener) {
        if (listener != null && mListeners.contains(listener) == false) {
            mListeners.add(listener);
        }
    }

    public boolean isProgressViewPosition(int position) {
        return shouldShowProgressView() && position == getCount() - 1;
    }

    private boolean shouldShowProgressView() {
        return mAdapter.getCount() > 0 && mCanReadMore;
    }

    public void canReadMore(boolean enable) {
        mCanReadMore  = enable;
    }

    public boolean isRefreshing() {
        return state.get() == REFRESHING_STATE;
    }

    public void handledRefresh() {
        assert(Looper.myLooper().equals(Looper.getMainLooper()));
        if (isRefreshing()) {
            mProgressView.setVisibility(View.GONE);
            state.set(NONE_STATE);
        }
    }

    public T getAdapter() {
        return mAdapter;
    }
}

File: library/src/main/kotlin/lab/prada/android/ui/lab/prada/android/ui/infinitescroll/InfiniteScrollAdapter.kt
package lab.prada.android.ui.lab.prada.android.ui.infinitescroll

import android.content.Context
import android.database.DataSetObserver
import android.os.Looper
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.Adapter
import android.widget.AbsListView
import android.widget.BaseAdapter
import android.widget.ProgressBar
import android.widget.RelativeLayout
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.atomic.AtomicInteger

/** Adds a progress row to a [BaseAdapter] and notifies listeners when it is displayed. */
class InfiniteScrollAdapter<T : BaseAdapter> : BaseAdapter {
    private val adapter: T
    private val progressView: View
    private val listeners = CopyOnWriteArrayList<InfiniteScrollListener>()
    private val state = AtomicInteger(NONE_STATE)

    @Volatile
    private var canReadMoreEnabled = true

    interface InfiniteScrollListener {
        fun onInfiniteScrolled()
    }

    constructor(context: Context, adapter: T, progressView: View) {
        this.adapter = adapter
        this.progressView = progressView
    }

    constructor(context: Context, adapter: T, itemWidth: Int, itemHeight: Int) {
        this.adapter = adapter
        this.progressView = RelativeLayout(context).apply {
            layoutParams = AbsListView.LayoutParams(itemWidth, itemHeight)
            gravity = Gravity.CENTER
            addView(ProgressBar(context))
        }
    }

    constructor(adapter: T, progressView: View) {
        this.adapter = adapter
        this.progressView = progressView
    }

    override fun areAllItemsEnabled(): Boolean = adapter.areAllItemsEnabled()

    override fun getDropDownView(
        position: Int,
        convertView: View?,
        parent: ViewGroup
    ): View? = if (isProgressViewPosition(position)) {
        null
    } else {
        adapter.getDropDownView(position, convertView, parent)
    }

    override fun getItemViewType(position: Int): Int = if (isProgressViewPosition(position)) {
        Adapter.IGNORE_ITEM_VIEW_TYPE
    } else {
        adapter.getItemViewType(position)
    }

    override fun getViewTypeCount(): Int = adapter.viewTypeCount

    override fun hasStableIds(): Boolean = adapter.hasStableIds()

    override fun isEmpty(): Boolean = adapter.isEmpty

    override fun registerDataSetObserver(observer: DataSetObserver) {
        adapter.registerDataSetObserver(observer)
    }

    override fun unregisterDataSetObserver(observer: DataSetObserver) {
        adapter.unregisterDataSetObserver(observer)
    }

    override fun getCount(): Int = if (shouldShowProgressView()) {
        adapter.count + 1
    } else {
        adapter.count
    }

    override fun getItem(position: Int): Any? = if (isProgressViewPosition(position)) {
        null
    } else {
        adapter.getItem(position)
    }

    override fun getItemId(position: Int): Long = if (isProgressViewPosition(position)) {
        -1L
    } else {
        adapter.getItemId(position)
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        if (!isProgressViewPosition(position)) {
            return adapter.getView(position, convertView, parent)
        }

        if (state.compareAndSet(NONE_STATE, REFRESHING_STATE)) {
            listeners.forEach(InfiniteScrollListener::onInfiniteScrolled)
        }

        if (progressView.visibility == View.GONE) {
            progressView.visibility = View.VISIBLE
        }
        return progressView
    }

    fun removeListener(listener: InfiniteScrollListener?) {
        if (listener != null) {
            listeners.remove(listener)
        }
    }

    fun addListener(listener: InfiniteScrollListener?) {
        if (listener != null) {
            listeners.addIfAbsent(listener)
        }
    }

    fun isProgressViewPosition(position: Int): Boolean =
        shouldShowProgressView() && position == count - 1

    fun canReadMore(enable: Boolean) {
        canReadMoreEnabled = enable
        notifyDataSetChanged()
    }

    fun isRefreshing(): Boolean = state.get() == REFRESHING_STATE

    fun handledRefresh() {
        check(Looper.myLooper() == Looper.getMainLooper()) {
            "handledRefresh() must be called on the main thread"
        }
        if (state.compareAndSet(REFRESHING_STATE, NONE_STATE)) {
            progressView.visibility = View.GONE
        }
    }

    fun getOriginalAdapter(): T = adapter

    fun getAdapter(): T = adapter

    private fun shouldShowProgressView(): Boolean = adapter.count > 0 && canReadMoreEnabled

    private companion object {
        const val NONE_STATE = 0
        const val REFRESHING_STATE = 1
    }
}
